packages/next/client/index.tsx

```diff
@@ -413,6 +415,7 @@ export async function initNext(opts: { webpackHMR?: any } = {}) {
     domainLocales,
     isPreview,
   })
+  publicRouterInstance = makePublicRouterInstance(router)
```

Comment:
We might need to tweak how `makePublicRouterInstance` is handling the fields to prevent this from being re-created on every render. Currently it's creating a new instance and copying the fields to prevent them from being treated as stateful.

Reply:
Doesn't putting it outside of render and creating it once at initialization, like this PR does, mitigate the issue?

Reply:
Oh, looks like we should go with https://github.com/vercel/next.js/pull/33875 approach instead? @devknoll

Reply:
It does but then the fields are never updated from the initial public router since they are copied here https://github.com/vercel/next.js/blob/39d3210776d37e7dc11e77b54a65b33b5454c0ce/packages/next/client/router.ts#L160-L170 and then the underlying router instance is updated but not the public router instance. So maybe instead of copying the fields we could Proxy them in development, erroring when setting to them to prevent them from being treated as stateful and in production we don't modify the router like this 🤔

Reply:
So, the issue is that _router_ is a bit of a misnomer here. When used globally, it means "routing functions + a reference to the currently visible route." When used locally (via a React hook), it means "routing functions + the currently rendered tree."

The router context _should_ change for each tree IMO. Consider something like this:

```jsx
function Foo() {
  const { query } = useRouter()
  return <span>{JSON.stringify(query)}</span>
}
```
1. If the context didn't change when `query` changed, how would we know to rerender?
2. If the value of `query` only pointed to the active route, how would we support `startTransition`, which renders the _next_ page concurrently in the background?

My approach in  #33875 is working on addressing this from this perspective, i.e. just creating a new public instance for each route/navigation, instead of generating a new one each time `AppContainer` is rerendered.